qiskit/dagcircuit/dagcircuit.py

```diff
@@ -610,14 +610,20 @@ def compose(self, other, edge_map=None, qubits=None, clbits=None, front=False, i
             return None
 
     def idle_wires(self):
-        """Return idle wires.
+        """Return idle wires (wires which only contain barriers are idle).
 
         Yields:
             Bit: Bit in idle wire.
         """
         for wire in self._wires:
+            barrier_counter = 0
             nodes = self.nodes_on_wire(wire, only_ops=False)
-            if len(list(nodes)) == 2:
+            length = 0
+            for x in list(nodes):
+                length += 1
+                if x.name == 'barrier':
+                    barrier_counter += 1
+            if length - barrier_counter == 2:
```

Comment:
This works, but it seems a bit "dirty" to me. Adding the `barrier_is_idle` flag (see https://github.com/Qiskit/qiskit-terra/pull/4481/files#r432798578) to this code might be hard. Adapting your approach, here is my suggestion:
```
            nodes = list(self.nodes_on_wire(wire, only_ops=False))
            if barrier_is_idle:
                nodes = [node for node in nodes if node.name is not 'barrier']
            if len(nodes) == 2:
                yield wire
```